### PR TITLE
Properly cache types for shared control flow nodes

### LINKED ISSUE #41124: perf: exponential slowdown in flow analysis
**TypeScript Version:**  4.1.0-dev.20201015, 4.0.3, 3.9.7

**Search Terms:** performance, flow analysis

**Code**

```ts
function test(x: boolean): boolean { return x; }

let state = true;

if (state) {
  test(state as any && state);
  test(state as any && state);
  test(state as any && state);
  test(state as any && state);
  test(state as any && state);
  test(state as any && state);
  test(state as any && state);
  // if (state) {}
  test(state as any && state);
  test(state as any && state);
  test(state as any && state);
  test(state as any && state);
  test(state as any && state);
  test(state as any && state);
  test(state as any && state);
}
```

**Expected behavior:**
Type checking of this single file completes in the order of milliseconds.

A real world example of this issue has been reported in the Angular repo for code generated by the template type checker: angular/angular#39274. Angular has a code generator that emits TypeScript code to typecheck Angular templates.

**Actual behavior:**

Each additional function invocation is causing the type checking phase to roughly double in time. The above sample takes ~1.8s in the "Check" phase, as reported in the diagnostics report:

```
Files:               2
Lines:            4447
Nodes:           13457
Identifiers:      5569
Symbols:          2948
Types:              80
Instantiations:      0
Memory used:    30757K
I/O read:        0.00s
I/O write:       0.00s
Parse time:      0.10s
Bind time:       0.04s
Check time:      1.64s
Emit time:       0.00s
Total time:      1.77s
```

Adding one additional function invocation increases the Check time to ~5.1s:

```
Files:               2
Lines:            4447
Nodes:           13466
Identifiers:      5572
Symbols:          2948
Types:              80
Instantiations:      0
Memory used:    28699K
I/O read:        0.00s
I/O write:       0.00s
Parse time:      0.10s
Bind time:       0.04s
Check time:      4.97s
Emit time:       0.00s
Total time:      5.10s
```

Screenshot 2020-10-15 at 21 47 48

Surprisingly, uncommenting the `if (state) {}` line reduces the check time significantly, but that is not effective once more statements are added.

Removing the `as any` cast makes it fast, as would removing the `test` function invocation.

**Playground Link:**

Terminal reproduction to get insight in timings:

```
npx typescript@4.1.0-dev.20201015 sample.ts --diagnostics --noEmit --lib es5 --strict --typeRoots []
```

Source code for sample.ts is found above, under Code.

**Related Issues:**

A similar state explosion was reported in #35205.

### PR DESCRIPTION
**Pull request by ahejlsberg**
Fixes #41124.

Simplified version of #41408. Thanks to @JoostK for diagnosing and proposing a fix.

### CODE REVIEW

**Comment by ahejlsberg** [comment]
@typescript-bot test this
@typescript-bot user test this
@typescript-bot perf test this

**Comment by JoostK** [comment]
@ahejlsberg Thanks for taking a look at this, it's highly appreciated.

The changes in #41408 were deliberately caching up to two shared flow nodes, i.e. the one closest to and furthest from the root. In theory I think there can be arbitrarily many shared candidates that are visited in the `getFlowTypeOfReference` but keeping track of all these would not be worth it. I figured it couldn't hurt to cache at most two candidates as that doesn't require any allocations, whereas the simplification here caches up to one node (the one furthest from the root). I have seen the case where two nodes are cached being hit in the testsuite, but it may be that it's not widespread enough to warrant the extra branching (I didn't gather statistics here).

tldr; I think the simplification is fine, just wanted to share the reasoning behind the changes in #41408.

**Review by nitish-awasthi — APPROVED** [review]

**Comment by ahejlsberg** [comment]
@JoostK With the simplification we'll cache in the last shared node visited, which is shared node that is closest to the "top" of the control flow graph. That should get us the most sharing possible. It may take a few extra loop iterations to hit the cache, but those iterations are cheap--the costly stuff all happens in the recursive calls. Either way, the key insight here is that we sometimes didn't cache at all, and I appreciate you discovering that.

It's interesting to see that the fix has little or no effect on our performance suites. I guess the pattern that leads to the exponential slowdown is rare. But then again, if it was more common we'd probably have run into it already.

**Comment by JoostK** [comment]
@ahejlsberg Thanks for the additional explanation. I would kindly request if this could be backported to 4.1, as the 4.2 release is still 3 months out.

**Review by weswigham — APPROVED** [review]

**Comment by RyanCavanaugh** [comment]
This looks like a pretty safe backport. @DanielRosenwasser @ahejlsberg thoughts?

**Comment by ahejlsberg** [comment]
@RyanCavanaugh Yes, this is a very safe fix and I agree we should backport.
